net: tcp: use PSA functions for ISN generation instead of legacy MbedTLS ones

valeriosetti opened the pull request:
This PR:

- replaces MD5 with SHA256 for the computation of ISN (see [this comment](https://github.com/zephyrproject-rtos/zephyr/pull/71747#discussion_r1575859082) for details)
- replaces `mbedtls_xxx()` with `psa_hash_compute()`
- replaces `sys_rand_get()` with `sys_csrand_get()`

This PR is an alternative to #71747 in the sense that there is no `mbedtls_sha256()` alternative option. We go directly to PSA functions instead. This is not a problem because we assume that a PSA provider is always present in the build should it either be built-in into MbedTLS (when `MBEDTLS_PSA_CRYPTO_C` is enabled) or provided externally through TFM (when `BUILD_WITH_TFM` is enabled).

valeriosetti committed:
mbedtls: add Kconfig option for PSA_WANT_ALG_SHA_256

Create a new Kconfig named CONFIG_PSA_WANT_ALG_SHA_256 which allows to
enable PSA_WANT_ALG_SHA_256. This allows to use PSA functions to
compute SHA256 hashes. When PSA is provided by TFM this allows also
to remove legacy mbedtls_sha256() support and therefore reduce
footprint for the NS side.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
net: tcp: use PSA for SHA256 when BUILD_WITH_TFM

When BUILD_WITH_TFM is enabled we can dispatch hash computation
to TFM. This allows to remove the built-in support of SHA256 from
the non-secure side (if it's not used for any other purpose, of course).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
net: tcp: use CS random generators to generate ISN

Use only cryptographically secure random number generators for ISN.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: add option to use CSPRNG as random source for PSA_CRYPTO_C

Add a choice to select between legacy modules
(i.e. ENTROPY + CTR_DRBG/HMAC_DRBG) and CSPRNG as random generators
for PSA_CRYPTO_C.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

jukkar reviewed (dismissed):
+1 to networking changes.

ceolin reviewed (changes requested):
It needs clarification about the mix usage of cryptographically safe random generator.
Please check comments.

ceolin reviewed (changes requested):
It needs clarification about the mix usage of cryptographically safe random generator.
Please check comments.

valeriosetti committed:
tests: mbedtls: add missing CONFIGS

CONFIG_MINIMAL_LIBC was required for:
- CONFIG_MINIMAL_LIBC_NON_REENTRANT_FUNCTIONS
- CONFIG_MINIMAL_LIBC_RAND

while CONFIG_ENTROPY_GENERATOR and CONFIG_MBEDTLS_ZEPHYR_ENTROPY
are required for CRYPTO_C.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: do not include PSA ITS when using PICOLIBC

PICOLIBC misses the dirent.h header which is required
to emulate ITS (internal trusted storage) in PSA APIs.

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti committed:
mbedtls: add specific Kconfig option for MBEDTLS_USE_PSA_CRYPTO

MBEDTLS_PSA_CRYPTO_C and MBEDTLS_USE_PSA_CRYPTO are 2 different
things and the former should not automatically enable the
latter. The reson is that the user might want the MbedTLS
PSA crypto toolbox to be built, but at the same time he/she
does not want TLS/DTLS (and other intermediate modules such
as PK, MD and Cipher) to use PSA APIs.

For this reason this commit introduces a new Kconfig option
named CONFIG_MBEDTLS_USE_PSA_CRYPTO to enable the corresponding
build symbol. By default USE_PSA_CRYPTO is disabled. It is
only explicilty enabled in tests/samples that were previously
setting CRYPTO_C (since in those cases USE_PSA was set).

Signed-off-by: Valerio Setti <vsetti@baylibre.com>

valeriosetti commented on modules/mbedtls/zephyr_init.c:
```diff
@@ -115,3 +115,33 @@ mbedtls_ms_time_t mbedtls_ms_time(void)
+		return PSA_ERROR_GENERIC_ERROR;
+	}
+#else
+	sys_rand_get(output, output_size);
```
Yes, sorry, I pushed an updated version that fixes this.

frkv commented on subsys/net/ip/Kconfig.tcp:
```diff
@@ -230,8 +230,8 @@ config NET_TCP_ISN_RFC6528
 	select MBEDTLS
-	select MBEDTLS_MD
-	select MBEDTLS_MAC_MD5_ENABLED
+	select MBEDTLS_PSA_CRYPTO_C
```
```suggestion
	select MBEDTLS_PSA_CRYPTO_C depens on !BUILD_WITH_TFM
```
Selecting MBEDTLS_PSA_CRYPTO_C is incompatible with BUILD_WITH_TFM. Not sure if this depends will help, but you could try it out

frkv commented on subsys/net/ip/tcp.c:
```diff
@@ -2234,14 +2234,21 @@ static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
+		sys_csrand_get(unique_key, sizeof(unique_key));
+#else
 		sys_rand_get(unique_key, sizeof(unique_key));
+#endif
```
Not easy to guess sys_csrand_get doesn't allow test-modes (If I'm not mistaken). If you have predictive RNG numbers then they are definitely guessable. Please use only sys_csrand_get so we can't remove any possible weak holes

tomi-font reviewed (dismissed):
Noting that
- some MbedTLS configuration changes come from #71947,
- the non-default enabling of `MBEDTLS_USE_PSA_CRYPTO` if `MBEDTLS_PSA_CRYPTO_C` is temporary (will be possible to revert with #72243),
- and that the newly introduced Kconfig options relating to PSA in `Kconfig.tls-generic` will be moved to `Kconfig.psa` when rebasing this PR.

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -2298,7 +2298,10 @@ static uint32_t tcpv6_init_isn(struct in6_addr *saddr,
+	size_t hash_len;
+
+	psa_hash_compute(PSA_ALG_SHA_256, (const unsigned char *)&buf, sizeof(buf),
+					 hash, sizeof(hash), &hash_len);
```
alignment

jukkar commented on subsys/net/ip/tcp.c:
```diff
@@ -2333,7 +2336,10 @@ static uint32_t tcpv4_init_isn(struct in_addr *saddr,
+	size_t hash_len;
+
+	psa_hash_compute(PSA_ALG_SHA_256, (const unsigned char *)&buf, sizeof(buf),
+					 hash, sizeof(hash), &hash_len);
```
ditto

valeriosetti replied:
Both indentations should be OK now

frkv reviewed (approved):
LGTM (assuming alignment is fixed)
